### DISCUSSION TITLE
Router violation with proprietary PDK

### DISCUSSION THREAD

**Discussion by sebinho** [The-OpenROAD-Project/OpenROAD#4987] [General]
Hi Guys,

I am posting here regarding an issue I am struggling with regarding the detailed routing of OR.
I am using a proprietary PDK, so it is difficult for me to share a complete example of the issue.

I am constantly running into metal spacing violations on M1 with my 45nm PDK. I systematically get the same error, typically on the same net(s). Because of that, the router always goes up to the max number of iterations (64) and never finds a solution.
Here is a snapshot of the issue at hand. 

![Screenshot from 2024-04-23 17-07-03](https://github.com/The-OpenROAD-Project/OpenROAD/assets/5883454/760a2c49-1fcb-4f10-aa12-ccccc09f12c9)

You can see on the bottom orange rectangle (this is a VIA between M1 and M2), that the metal spacing is 0.06um. This violates the metal spacing rule of 0.07. M1 is in dark blue. You can see M2 in cyan in the picture below.

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/5883454/6a93f3e6-8420-4407-8a70-54afad90ffb2)
 
What I noticed is that the routing tool is able to select between 2 types of vias (on its own). Sometimes he selects a horizontal via (like the one that is problematic) and sometimes he takes a vertical via (like the one show in the upper part of the image where the 0.07um rule is respected).
The solution to the problem is quite straightforward when done manually. Simply replace the problematic via with its vertical version. But somehow the tool cannot see that.

Is there something that can be done for such an issue @maliberty ? I understand that it is not easy to debug as I cannot share my PDK . Apologies for that.
Thanks for your help.

**Comment by maliberty** [comment]
You said to goes to 64 iterations.  Does drt report these violation itself or are they only seen by another drc tool?  

Precision Innovations does provide paid support under NDA if that makes sense for you.

@osamahammad21 any thoughts?

**Reply by sebinho** [reply]
This is actually a violation reported by OpenROAD and also by our DRC proprietary tool. 
regarding paid services, we are not looking at this track for the moment, we are first trying to find a solution and then we will see what we can do next. 
Worse case those violations (always the same ones) can be changed manually provided that there are not too many

**Comment by maliberty** [comment]
Does the drc straddle the boundary of two different cells?  (ie the pins are in different instances)

**Reply by sebinho** [reply]
I just checked, and in this case it seems to be indeed right at the boundary between 2 different cells (a FF and a MUX, see below).
I will rerun the tool with different settings to see if it always occurs on a boundary and let you know.

![via_issue](https://github.com/The-OpenROAD-Project/OpenROAD/assets/5883454/129234bc-29f2-433e-beee-af3c31370347)

**Comment by sebinho** [comment]
I reran the tool with different settings and 1 get 1 violation for metal spacing. It is a different location in my design but once again it seems to be at the boundary between 2 different cells. See below red line for separation.

![via_issue2](https://github.com/The-OpenROAD-Project/OpenROAD/assets/5883454/6e9d737f-dd9b-4bf7-aa9e-4184a338228b)

Any clue how this could be solved?
Thanks a lot

**Reply by maliberty** [reply]
One idea would be to use -via_in_pin_bottom_layer / -via_in_pin_top_layer to force the via to fit inside the pin shape.   You should set placement padding between the cells but that would cost density.  Otherwise I think we would need a test case.

**Reply by sebinho** [reply]
Thanks for the suggestions.
Would you mind telling me what those via_in_pin_bottom_layer/via_in_pin_top_layer settings mean? I can’t find any documentation about it.

**Reply by maliberty** [reply]
The specify a layer range (generally just the std cell pin layer) for which the via must be fully enclosed by the pin to be legal.

**Reply by sebinho** [reply]
Thanks 🙏 🙏

**Reply by sebinho** [reply]
@maliberty I did try to use the `-via_in_pin_bottom_layer / -via_in_pin_top_layer` option and I did set them both to M1 (this is the STD cells layer) but it did not change the issue I have (metal spacing violation).
Could you maybe explain how I can try to add padding between the cells?

**Reply by maliberty** [reply]
That sounds odd as the vias should be fully enclosed.   In ORFS you can set CELL_PAD_IN_SITES_GLOBAL_PLACEMENT and CELL_PAD_IN_SITES_DETAIL_PLACEMENT (say =1).

**Reply by sebinho** [reply]
This is what I added to my python script:
```
chip.set('tool', 'openroad', 'task', 'route', 'var', 'via_in_pin_bottom_layer', 'M1')
chip.set('tool', 'openroad', 'task', 'route', 'var', 'via_in_pin_top_layer', 'M1')
```

After many iterations, I get the following error:
```
violation type: Metal Spacing
srcs: net:apb_psel_o net:vdd 
bbox = ( 49.485, 60.025 ) - ( 49.55, 60.095 ) on Layer M1
```

And the snapshot (again at the edge of a flip-flop, incorrect via used when a vertical one would have worked). Issue is the via at the top right.

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/5883454/2addd15f-dc55-4b79-9b44-d0d8e554da51)

**Reply by maliberty** [reply]
I don't know what python script that is.  Is it Silicon Compiler?  I wonder if that is actually reaching OR.

**Reply by sebinho** [reply]
sorry, my mistake. It is indeed SiliconCompiler.
I checked the internal scripts used by SC and it was indeed not reached due to the fact that this is not yet used by anybody. I was able to set it properly by tweaking it a bit.

I am not getting any violations anymore, thanks a lot. 
The router still needs around 50 iterations to finish, so it is quite slow especially for a very small test design. But it might be related to the fact that my PDK has many more rules to satisfy than what we have for example with Sky130. Not sure just wild guessing.

Really appreciate your help on this matter

**Reply by maliberty** [reply]
Via in pin does make routing more difficult due to restricting the pin access.   With more pdk access we could probably find a better solution but at least this works.  Lowering the placement density could help make things faster if you have the space.

@gadfort see the discussion above about SiC

**Reply by sebinho** [reply]
Thanks for the suggestions.
Regarding SC @gadfort, I did not have to change `sc_route.scl`, but I had to use the following names for the OpenROAD variables instead (add `drt_` prefix):
```
chip.set('tool', 'openroad', 'task', 'route', 'var', 'drt_via_in_pin_bottom_layer', 'M1')
chip.set('tool', 'openroad', 'task', 'route', 'var', 'drt_via_in_pin_top_layer', 'M1')
```

It would also be nice to add some information in the helper of those variables, as for the time being it is set to "to do".

**Reply by gadfort** [reply]
@sebinho I believe the settings should be applied to: `drt_via_in_pin_bottom_layer` and `drt_via_in_pin_top_layer`. The picture above doesn't look like there is a density issue if it's access within a standard cell.

**Reply by sebinho** [reply]
@gadfort, indeed, I guess we replied at the same time. Thanks for commenting

**Reply by gadfort** [reply]
I need @maliberty to provide definitions to those unless the online docs finally have them properly defined. (from the flag itself it's not clear why there is a top and bottom layer.)

**Reply by gadfort** [reply]
@maliberty this doesn't actually explain what the flag does: https://openroad.readthedocs.io/en/latest/main/src/drt/README.html#options

**Reply by maliberty** [reply]
@osamahammad21 please improve the docs for the via in pin option

**Comment by sebinho** [comment]
@maliberty do you have any recommendations on where to look first in order to speed up the routing for a proprietary PDK?
Now that I added an extra layer of complexity (custom DEF file to specify core/die size and pins locations, custom power grid), it often crash due to not being able to access pins. Or it takes forever to eventually not solve all violations. 
Should I incrementally try to simplify the technology (.LEF) file? Any other things I should look into?
My test design (very small module) goes through Sky130/freepdk45 in a matter of minutes even with a very similar DEF file and power grid…
Thanks for your help

**Reply by sebinho** [reply]
One of the things I did for the time being in order to remove the « cannot find route to pin » is to tell the tool in my stackup spec not to use metal 1 layer for routing…

these issues do not occur if I let the tool place the pins randomly around my die area…at least for the pin access, routing still takes a long time

**Reply by maliberty** [reply]
You can use -droute_end_iter <iter> to make it stop earlier if it is just stuck on a few DRCs that you plan to fix manually.  You can look at the reports at the end of the iteration to get a sense of the sort of violation it having trouble with if plan to simplify.

In general the sort of rules matter a lot.  Pin access is tricky to diagnose with no data.  Its more of a developer tool but you are welcome to try `detailed_route_debug -pa -pa_markers -pin <inst_name>:<pin_name>` to get more details on a pin access problem.  You give that before calling detailed_route while running in the GUI and it will show you how it is trying to access the pin.

**Reply by sebinho** [reply]
Thanks for the tips. I will try it out and let you know

**Reply by sebinho** [reply]
I did a lot of various testing. 
Regarding the "pin access" issues, I tried the developer option with `detailed_route_debug`. This is indeed not trivial and hard to figure out what is going wrong.

What I did is drastically simplify the technology LEF file, knowing that I would get tons of DRC violations. After doing this, no more pin access issues and the router is extremely fast (like a 3-5 iterations).

Then I can add some rules 1 by 1 in the technology file, and see when it starts failing but it quickly does.
At first I thought it could be related to my power grid setup or even my custom pin placement (as changing them could have an impact on the errors received), but now I think it all boils down to the rules defined in the LEF file.

This is very tricky to debug and to find an optimum. I suppose there isn't a miracle solution. Or maybe there is something I don't grasp.

Should I have the assumption that it is possible to port any rules deck with OpenROAD?
Not sure if you have any other suggestions or what would be the best approach to have when porting a new PDK with OpenROAD?
Thanks @maliberty

**Reply by sebinho** [reply]
I can share one example of a rule that the openROAD router does not seem to apply properly.
I have a rule defined in one of my VIAs. This is a rule that must be applied when a via array is used.  Here is the rule in question:
```
SPACING 0.34 ;                                           
SPACING 0.54 ADJACENTCUTS 3 WITHIN 0.56 ;
```
When we have simple vias, they will have a spacing of 0.34um. But when we have a more complex via array, it should use a spacing of 0.54um. This does not work and I can clearly see in the GDS that the spacing is kept at 0.34.

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/5883454/a9a9b7b2-9df1-4f9c-b875-fc409ca1082b)

My current workaround is to set the first line 0.54 instead of 0.34 as I know those layers are only used for power routing and I can just use vias with a bigger spacing.
This issue concerns the upper layers (thick ones).

**Comment by maliberty** [comment]
> Should I have the assumption that it is possible to port any rules deck with OpenROAD?

That's probably too strong a statement (eg we don't support all rules below 12nm) but for older nodes it should be possible.  More rules certainly will slow down the router if they make it more difficult to find legal solutions.

@osamahammad21 am I correct in thinkin we support ADJACENTCUTS / WITHIN?

**Reply by sebinho** [reply]
I can also comment about another issue I noticed. It is related to the insertion of multiple vias using the `MINIMUMCUT` rule.
On my metal1 layer, I have several `MINIMUMCUT` rules.

When I use the following rules (subset of the official rules), the router works properly and goes to the end.
```
MINIMUMCUT 2 WIDTH 0.210 WITHIN 0.14 ;
MINIMUMCUT 4 WIDTH 0.550 WITHIN 0.14 ;
```

But when I use the following extended rules, the router fails immediately with a bunch of errors of type `no access point for _xxx_/Z`.
```
MINIMUMCUT 2 WIDTH 0.210 WITHIN 0.14 ;
MINIMUMCUT 4 WIDTH 0.550 WITHIN 0.14 ;
MINIMUMCUT 2 WIDTH 0.21 FROMABOVE LENGTH 0.21 WITHIN 1.141 ;
```
It is not related to the addition of `FROMABOVE` as I get the same errors if I add this line instead of the previously added line:
`MINIMUMCUT 2 WIDTH 0.21 LENGTH 0.21 WITHIN 1.141 ;`

Any idea?

**Reply by maliberty** [reply]
We do not currently support putting in multi-cut vias as they come with a lot more drc rule requirements.

**Reply by sebinho** [reply]
thanks for letting me know, this makes sense that I cannot satisfy this rule.

**Reply by sebinho** [reply]
After lots of tickering, we do have something in place with our proprietary PDK that runs relatively fast but we are unable to be completely DRC clean. 
Even if we discard the DRC rules regarding multi-cut vias, there are still a couple of rules that seem not to be understood by the tool. I am not sure we can do much better. 
I would be curious to see the technology file used in a real 45nm PDK. 

@maliberty Do you know if any tapeout was ever made with a 45nm? I have seen lower nodes being used with OpenROAD like GF15. Is there any publicly available 45nm technology file aside from the freepdk45 one? This one is way too simple and does not have many DRC rules. It would be a good reference to compare to.

**Reply by maliberty** [reply]
I don't know of any public manufacturable 45nm PDK.  I think we are reaching the limits of what we can do without a support contract and an NDA.

**Reply by sebinho** [reply]
Makes sense, understood.
Anyways much appreciated your help.
If we want to take this further at some point, should we reach out to you from your website?

**Reply by maliberty** [reply]
Yes
